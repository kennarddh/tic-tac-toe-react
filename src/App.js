import React, { useState, useCallback, useEffect } from 'react'

// Components
import Square from 'Components/Square/Square'

// Styles
import {
	Container,
	GameContainer,
	RestartButton,
	UndoRedoButton,
	WinnerText,
} from './Styles'

const App = () => {
	const [Squares, SetSquares] = useState(Array(9).fill(null))
	const [History, SetHistory] = useState([])
	const [IsNowX, SetIsNowX] = useState(true)
	const [IsCanRestart, SetIsCanRestart] = useState(false)
	const [Winner, SetWinner] = useState(null)
	const [IsDraw, SetIsDraw] = useState(false)

	const OnClick = useCallback(
		index => {
			if (Squares[index] !== null) return
			if (Winner) return
			if (IsDraw) return

			SetSquares(squares => {
				const squaresClone = squares.slice()

				squaresClone[index] = IsNowX ? 'X' : 'O'

				return squaresClone
			})

			SetIsNowX(isNowX => {
				return !isNowX
			})

			SetHistory(history => [...history, Squares])
		},
		[IsNowX, Winner, History]
	)

	const CalculateWinner = squares => {
		const lines = [
			[0, 1, 2],
			[3, 4, 5],
			[6, 7, 8],
			[0, 3, 6],
			[1, 4, 7],
			[2, 5, 8],
			[0, 4, 8],
			[2, 4, 6],
		]

		for (let i = 0; i < lines.length; i++) {
			const [a, b, c] = lines[i]

			if (
				squares[a] &&
				squares[a] === squares[b] &&
				squares[a] === squares[c]
			) {
				return squares[a]
			}
		}

		return null
	}

	const Restart = useCallback(() => {
		SetSquares(Array(9).fill(null))
		SetIsNowX(true)
		SetIsCanRestart(false)
		SetWinner(null)
		SetHistory([])
		SetIsDraw(false)
	}, [])

	const Undo = () => {
		if (Winner) return
		if (IsDraw) return
		if (History.length === 0) return

		SetSquares(History[History.length - 1])

		SetHistory(history => history.slice(0, history.length - 1))
	}

	useEffect(() => {
		const winner = CalculateWinner(Squares)

		if (winner) {
			SetIsCanRestart(true)
		} else {
			const isDraw = Squares.every(square => square !== null)

			if (isDraw) {
				SetIsCanRestart(true)
			}

			SetIsDraw(isDraw)
		}

		SetWinner(winner)
	}, [Squares])

	return (
		<Container>
			<h1>Tic Tac Toe</h1>

			<p>The game is played on a 3x3 grid.</p>

			<p>The game is played by two players.</p>

			<p>The first player to get three in a row wins.</p>

			<UndoRedoButton onClick={Undo} data-testid='undo-button'>
				Undo
			</UndoRedoButton>

			<WinnerText data-testid='result' showing={IsDraw || Winner}>
				{IsDraw ? 'Draw' : Winner ? `Winner: ${Winner}` : ''}
			</WinnerText>

			<GameContainer>
				{Squares.map((square, index) => (
					<Square
						key={index}
						index={index}
						value={square}
						onClick={OnClick}
					/>
				))}
			</GameContainer>

			<RestartButton
				showing={IsCanRestart}
				onClick={Restart}
				data-testid='restart-button'
			>
				Play Again?
			</RestartButton>

			<p>
				Github:{' '}
				<a href='https://github.com/kennarddh/tic-tac-toe-react'>
					Tic Tac Toe React
				</a>
			</p>
		</Container>
	)
}

export default App
