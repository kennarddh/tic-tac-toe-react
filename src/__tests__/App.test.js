import React from 'react'
import App from '../App'

import { render } from '@testing-library/react'
import userEvent from '@testing-library/user-event'

import 'jest-styled-components'

describe('This suit is to test the App component', () => {
	test('Snapshot of App', () => {
		const { asFragment } = render(<App />)

		expect(asFragment()).toMatchSnapshot()
	})

	test('All square rendered with value', () => {
		const { getByTestId } = render(<App />)

		for (let i = 0; i < 9; i++) {
			expect(getByTestId(`square-${i}`)).toHaveTextContent('')
		}
	})

	test('Click on square', () => {
		const { getByTestId } = render(<App />)

		userEvent.click(getByTestId('square-0'))

		expect(getByTestId('square-0')).toHaveTextContent('X')
	})

	test('Click on square and check if all square rendered with no value', () => {
		const { getByTestId } = render(<App />)

		for (let i = 0; i < 9; i++) {
			expect(getByTestId(`square-${i}`)).toHaveTextContent('')
		}
	})

	test('Taking turns and click all square', () => {
		const { getByTestId } = render(<App />)

		userEvent.click(getByTestId('square-0'))
		expect(getByTestId('square-0')).toHaveTextContent('X')

		userEvent.click(getByTestId('square-1'))
		expect(getByTestId('square-1')).toHaveTextContent('O')

		userEvent.click(getByTestId('square-2'))
		expect(getByTestId('square-2')).toHaveTextContent('X')

		userEvent.click(getByTestId('square-4'))
		expect(getByTestId('square-4')).toHaveTextContent('O')

		userEvent.click(getByTestId('square-3'))
		expect(getByTestId('square-3')).toHaveTextContent('X')

		userEvent.click(getByTestId('square-5'))
		expect(getByTestId('square-5')).toHaveTextContent('O')

		userEvent.click(getByTestId('square-7'))
		expect(getByTestId('square-7')).toHaveTextContent('X')

		userEvent.click(getByTestId('square-6'))
		expect(getByTestId('square-6')).toHaveTextContent('O')

		userEvent.click(getByTestId('square-8'))
		expect(getByTestId('square-8')).toHaveTextContent('X')
	})

	test('Check X win', () => {
		const { getByTestId } = render(<App />)

		expect(getByTestId('result')).toHaveTextContent('')

		userEvent.click(getByTestId('square-0'))
		expect(getByTestId('square-0')).toHaveTextContent('X')

		userEvent.click(getByTestId('square-1'))
		expect(getByTestId('square-1')).toHaveTextContent('O')

		userEvent.click(getByTestId('square-2'))
		expect(getByTestId('square-2')).toHaveTextContent('X')

		userEvent.click(getByTestId('square-3'))
		expect(getByTestId('square-3')).toHaveTextContent('O')

		userEvent.click(getByTestId('square-4'))
		expect(getByTestId('square-4')).toHaveTextContent('X')

		userEvent.click(getByTestId('square-5'))
		expect(getByTestId('square-5')).toHaveTextContent('O')

		userEvent.click(getByTestId('square-6'))
		expect(getByTestId('square-6')).toHaveTextContent('X')

		expect(getByTestId('result')).toHaveTextContent('Winner: X')
	})

	test('Check O win', () => {
		const { getByTestId } = render(<App />)

		expect(getByTestId('result')).toHaveTextContent('')

		userEvent.click(getByTestId('square-0'))
		expect(getByTestId('square-0')).toHaveTextContent('X')

		userEvent.click(getByTestId('square-1'))
		expect(getByTestId('square-1')).toHaveTextContent('O')

		userEvent.click(getByTestId('square-3'))
		expect(getByTestId('square-3')).toHaveTextContent('X')

		userEvent.click(getByTestId('square-4'))
		expect(getByTestId('square-4')).toHaveTextContent('O')

		userEvent.click(getByTestId('square-2'))
		expect(getByTestId('square-2')).toHaveTextContent('X')

		userEvent.click(getByTestId('square-7'))
		expect(getByTestId('square-7')).toHaveTextContent('O')

		expect(getByTestId('result')).toHaveTextContent('Winner: O')
	})

	test('Check restart game', () => {
		const { getByTestId } = render(<App />)

		expect(getByTestId('restart-button')).toHaveStyle(`visibility: hidden`)

		userEvent.click(getByTestId('square-0'))
		expect(getByTestId('square-0')).toHaveTextContent('X')

		userEvent.click(getByTestId('square-1'))
		expect(getByTestId('square-1')).toHaveTextContent('O')

		userEvent.click(getByTestId('square-3'))
		expect(getByTestId('square-3')).toHaveTextContent('X')

		userEvent.click(getByTestId('square-4'))
		expect(getByTestId('square-4')).toHaveTextContent('O')

		userEvent.click(getByTestId('square-2'))
		expect(getByTestId('square-2')).toHaveTextContent('X')

		userEvent.click(getByTestId('square-7'))
		expect(getByTestId('square-7')).toHaveTextContent('O')

		expect(getByTestId('restart-button')).toHaveStyle(`visibility: visible`)

		userEvent.click(getByTestId('restart-button'))

		expect(getByTestId('restart-button')).toHaveStyle(`visibility: hidden`)

		for (let i = 0; i < 9; i++) {
			expect(getByTestId(`square-${i}`)).toHaveTextContent('')
		}

		expect(getByTestId('result')).toHaveTextContent('')
		expect(getByTestId('result')).toHaveStyle(`visibility: hidden`)

		userEvent.click(getByTestId('square-0'))
		expect(getByTestId('square-0')).toHaveTextContent('X')
	})

	test('Check undo button', () => {
		const { getByTestId } = render(<App />)

		userEvent.click(getByTestId('square-0'))
		expect(getByTestId('square-0')).toHaveTextContent('X')

		userEvent.click(getByTestId('undo-button'))

		for (let i = 0; i < 9; i++) {
			expect(getByTestId(`square-${i}`)).toHaveTextContent('')
		}

		userEvent.click(getByTestId('undo-button'))

		for (let i = 0; i < 9; i++) {
			expect(getByTestId(`square-${i}`)).toHaveTextContent('')
		}
	})

	test("Don't undo if game finished", () => {
		const { getByTestId } = render(<App />)

		userEvent.click(getByTestId('square-0'))
		expect(getByTestId('square-0')).toHaveTextContent('X')

		userEvent.click(getByTestId('square-1'))
		expect(getByTestId('square-1')).toHaveTextContent('O')

		userEvent.click(getByTestId('square-3'))
		expect(getByTestId('square-3')).toHaveTextContent('X')

		userEvent.click(getByTestId('square-4'))
		expect(getByTestId('square-4')).toHaveTextContent('O')

		userEvent.click(getByTestId('square-2'))
		expect(getByTestId('square-2')).toHaveTextContent('X')

		userEvent.click(getByTestId('square-7'))
		expect(getByTestId('square-7')).toHaveTextContent('O')

		userEvent.click(getByTestId('undo-button'))

		expect(getByTestId('square-0')).toHaveTextContent('X')
		expect(getByTestId('square-1')).toHaveTextContent('O')
		expect(getByTestId('square-3')).toHaveTextContent('X')
		expect(getByTestId('square-4')).toHaveTextContent('O')
		expect(getByTestId('square-2')).toHaveTextContent('X')
		expect(getByTestId('square-7')).toHaveTextContent('O')
	})

	test('Check draw', () => {
		const { getByTestId } = render(<App />)

		expect(getByTestId('result')).toHaveTextContent('')

		userEvent.click(getByTestId('square-0'))
		expect(getByTestId('square-0')).toHaveTextContent('X')

		userEvent.click(getByTestId('square-1'))
		expect(getByTestId('square-1')).toHaveTextContent('O')

		userEvent.click(getByTestId('square-2'))
		expect(getByTestId('square-2')).toHaveTextContent('X')

		userEvent.click(getByTestId('square-4'))
		expect(getByTestId('square-4')).toHaveTextContent('O')

		userEvent.click(getByTestId('square-3'))
		expect(getByTestId('square-3')).toHaveTextContent('X')

		userEvent.click(getByTestId('square-5'))
		expect(getByTestId('square-5')).toHaveTextContent('O')

		userEvent.click(getByTestId('square-7'))
		expect(getByTestId('square-7')).toHaveTextContent('X')

		userEvent.click(getByTestId('square-6'))
		expect(getByTestId('square-6')).toHaveTextContent('O')

		userEvent.click(getByTestId('square-8'))
		expect(getByTestId('square-8')).toHaveTextContent('X')

		expect(getByTestId('result')).toHaveTextContent('Draw')
	})

	test('Check result hidden', () => {
		const { getByTestId } = render(<App />)

		expect(getByTestId('result')).toHaveTextContent('')
		expect(getByTestId('result')).toHaveStyle(`visibility: hidden`)

		userEvent.click(getByTestId('square-0'))
		expect(getByTestId('square-0')).toHaveTextContent('X')

		userEvent.click(getByTestId('square-1'))
		expect(getByTestId('square-1')).toHaveTextContent('O')

		userEvent.click(getByTestId('square-2'))
		expect(getByTestId('square-2')).toHaveTextContent('X')

		userEvent.click(getByTestId('square-4'))
		expect(getByTestId('square-4')).toHaveTextContent('O')

		userEvent.click(getByTestId('square-3'))
		expect(getByTestId('square-3')).toHaveTextContent('X')

		userEvent.click(getByTestId('square-5'))
		expect(getByTestId('square-5')).toHaveTextContent('O')

		userEvent.click(getByTestId('square-7'))
		expect(getByTestId('square-7')).toHaveTextContent('X')

		userEvent.click(getByTestId('square-6'))
		expect(getByTestId('square-6')).toHaveTextContent('O')

		userEvent.click(getByTestId('square-8'))
		expect(getByTestId('square-8')).toHaveTextContent('X')

		expect(getByTestId('result')).toHaveStyle(`visibility: visible`)
	})
})
