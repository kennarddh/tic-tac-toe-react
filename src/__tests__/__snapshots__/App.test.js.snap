// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`This suit is to test the App component Snapshot of App 1`] = `
<DocumentFragment>
  .c4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  padding: 10px;
  width: 100px;
  height: 100px;
  background-color: #f5f5f5;
  box-sizing: border-box;
  border: 1px solid #ccc;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  padding: 10px;
  width: 100vw;
  height: 100vh;
  background-color: #f5f5f5;
  box-sizing: border-box;
}

.c0 h1 {
  margin: 10px 0;
}

.c0 p {
  margin: 5px 0;
}

.c3 {
  margin-top: 20px;
  display: grid;
  grid-template-columns: repeat(3,1fr);
  grid-template-rows: repeat(3,1fr);
  width: 300px;
  height: 300px;
}

.c5 {
  margin-top: 20px;
  min-width: 100px;
  height: 30px;
  background-color: #f5f5f5;
  border: 1px solid #ccc;
  border-radius: 5px;
  cursor: pointer;
  outline: none;
  font-size: 16px;
  font-weight: bold;
  color: #333;
  -webkit-transition: all 0.2s ease-in-out;
  transition: all 0.2s ease-in-out;
  visibility: hidden;
}

.c5:hover {
  background-color: #ccc;
  color: #333;
}

.c1 {
  min-width: 100px;
  height: 30px;
  background-color: #f5f5f5;
  border: 1px solid #ccc;
  border-radius: 5px;
  cursor: pointer;
  outline: none;
  font-size: 16px;
  font-weight: bold;
  color: #333;
  -webkit-transition: all 0.2s ease-in-out;
  transition: all 0.2s ease-in-out;
}

.c1:hover {
  background-color: #ccc;
  color: #333;
}

.c2 {
  visibility: hidden;
}

<div
    class="c0"
  >
    <h1>
      Tic Tac Toe
    </h1>
    <p>
      The game is played on a 3x3 grid.
    </p>
    <p>
      The game is played by two players.
    </p>
    <p>
      The first player to get three in a row wins.
    </p>
    <button
      class="c1"
      data-testid="undo-button"
    >
      Undo
    </button>
    <p
      class="c2"
      data-testid="result"
    />
    <div
      class="c3"
    >
      <div
        class="c4"
        data-testid="square-0"
      />
      <div
        class="c4"
        data-testid="square-1"
      />
      <div
        class="c4"
        data-testid="square-2"
      />
      <div
        class="c4"
        data-testid="square-3"
      />
      <div
        class="c4"
        data-testid="square-4"
      />
      <div
        class="c4"
        data-testid="square-5"
      />
      <div
        class="c4"
        data-testid="square-6"
      />
      <div
        class="c4"
        data-testid="square-7"
      />
      <div
        class="c4"
        data-testid="square-8"
      />
    </div>
    <button
      class="c5"
      data-testid="restart-button"
    >
      Play Again?
    </button>
    <p>
      Github: 
      <a
        href="https://github.com/kennarddh/tic-tac-toe-react"
      >
        Tic Tac Toe React
      </a>
    </p>
  </div>
</DocumentFragment>
`;
